### PR TITLE
docs: draft BootstrapPeers IER

### LINKED ISSUE #101: Decide on implementation of caught-up-ness information needed by Network
There are two general ways to implement the functionality described in #91:

 - Implement a minimal version of the interface, likely based on some recency condition of the immutable/volatile tip.
 - Decide that the functionality nicely fits into the Genesis State Machine, and we should implement the full/non-trivial parts of the GSM already today.

For context, the Genesis State Machine (GSM) is a part of the current Genesis design that determines when to "disable" certain properties/functionalities of the Consensus layer, as part of "switching" from Genesis to Praos. However, it does not depend on other Genesis components and could be implemented today.

The goal of this ticket is to decide on the implementation route we want to take.

---

Some notes from the 1:1 with @nfrisby today:

 - In general, it seems feasible to use the GSM for that purpose; the core question is, in both cases: "Can we safely resume operation in Praos mode, or do we have to switch to a (in some way more expensive) Syncing mode?"
    - In particular, the transition rule "All peers claim no more headers" for "Syncing -> CaughtUp" also works without Genesis, as we can/have to assume that we are only connected to trusted peers (probably just one) during Syncing.
    - We should avoid transient intermediary states, such as entering the CaughtUp state on startup just to immediately leave it due to our tip being too old.
 - The current interface in https://github.com/input-output-hk/ouroboros-network/pull/4555 does not allow Consensus to "notify" the Networking layer that it should switch back to Syncing after we were caught up, as Network decides when to call the function, implemented by Consensus, that returns the ledger peers to use.
   
   But that might actually be fine, if the transition "CaughtUp -> Syncing" always involves losing all of our peers (in which case Network has to call the function). That should e.g. true when "closing the laptop lid", but also when falling behind due to CPU starvation, as our peers will disconnect from us when the fork gets too deep. We should check this in more detail.

### PR DESCRIPTION
**Pull request by nfrisby**
Closes #101.

### CODE REVIEW

**Review by bolt12 — COMMENTED** [review]
This looks good to me :) I think we can settle the interface to be `lpGetPeers :: STM m (SlotNo, [(PoolStake, NonEmpty RelayAccessPoint)])`, would you agree? If yes then I have a question:

What's the relation or expected interaction between `State` and `lpGetPeers`? I ask this because we will have both `UseLedgerAfter` and `UseLedgerCloseToTip` (for now at least), the former requires a `SlotNo` and the latter does not. However, I think that if `UseLedgerCloseToTip` is enabled and `State` is `CaughtUp` we should get an empty list, but if `UseLedgerAfter` is enabled I think we shouldn't get an empty list. What do you think?

File: docs/website/docs/bootstrap-peers-IER.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  Delays here would directly threaten the security of the node.
+
+- `OnlyBootstrap -> CaughtUp` should be prompt so that the centralized, relatively-few bootstrap peers are relieved of load as soon as possible.
+  Delays here would no directly threaten the security of the node.
+  However, wasting the centralized resources would threaten the ability of nodes to join the net, ie the availability of the whole net.
+
+Desideratum 3.
+The node should not return to `OnlyBootstrap` every time it restarts/briefly loses network/etc.
+Such unnecessary connections would also put unnecessary load on the centralized, relatively-few bootstrap peers.
+This is the point of the "Node start-up" psuedo state in the diagram above.
```
Comment (amesgen) [inline docs/website/docs/bootstrap-peers-IER.md:42]:
```suggestion
This is the point of the "Node start-up" pseudo state in the diagram above.
```
:smile:

File: docs/website/docs/bootstrap-peers-IER.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+- For the sake of efficient reuse, the Diffusion Layer must manage the disconnections from all peers upon the `CaughtUp -> OnlyBootstrap` transition.
+  It would be wasteful for the Consensus Layer to achieve this less carefully, eg by simply killing those threads.
+
+  Therefore, the Consensus Layer needs to promptly notify the Diffusion Layer when the node should make the `CaughtUp -> OnlyBootstrap` transition.
+  Ideally, this happens immediately, so the implementation will likely involve an `STM` `TVar`.
+
+  As the volatile tip age approaches X, the Consensus Layer could forewarn the Diffusion Layer, eg "it seems like the transition back to OnlyBootstrap will be necessary soon; please prepare", if that would be helpful.
+
+- For similar reasons, the Diffusion Layer should also manage the disconections from all peers upon the `OnlyBootstrap -> CaughtUp` transition.
```
Comment (amesgen) [inline docs/website/docs/bootstrap-peers-IER.md:60]:
```suggestion
- For similar reasons, the Diffusion Layer should also manage the disconections from all peers (usually, just one bootstrap peer) upon the `OnlyBootstrap -> CaughtUp` transition.
```
This also reminds of whether this document should mention these BlockFetch-only peers as potential future work?

File: docs/website/docs/bootstrap-peers-IER.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+Desideratum 1.
+The following state machine depicts the desired behavior of the node.
+
+```mermaid
+
+graph    
+    OnlyBootstrap[OnlyBootstrap]
+    CaughtUp[CaughtUp]
+
+    OnlyBootstrap -- "no peers claim to have\nsubsequent headers" --> CaughtUp
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply docs/website/docs/bootstrap-peers-IER.md:17]: Yeah, we should add that part here.

**Review by dnadales — COMMENTED** [review]
Could we put this in a ADR, [eg](https://github.com/input-output-hk/cardano-ledger/blob/master/docs/adr/2022-12-05_002-version-serialization-schemes.md). If so we might want to start off with [such a doc](https://github.com/input-output-hk/cardano-ledger/blob/master/docs/adr/2022-12-01_001-record-architectural-decisions.md).

File: docs/website/docs/bootstrap-peers-IER.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/docs/bootstrap-peers-IER.md]: I phrased it as
> no peers claim to have subsequent headers, and node has finished chain selection

File: docs/website/docs/bootstrap-peers-IER.md (hunk 2 above)
Comment (amesgen) [reply docs/website/docs/bootstrap-peers-IER.md]: I did not go with mentioning the BlockFetch-only peers (https://github.com/input-output-hk/ouroboros-consensus/issues/189), felt sufficiently orthogonal.

File: docs/website/docs/bootstrap-peers-IER.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  Ideally, this happens immediately, so the implementation will likely involve an `STM` `TVar`.
+
+  As the volatile tip age approaches X, the Consensus Layer could forewarn the Diffusion Layer, eg "it seems like the transition back to OnlyBootstrap will be necessary soon; please prepare", if that would be helpful.
+
+- For similar reasons, the Diffusion Layer should also manage the disconections from all peers upon the `OnlyBootstrap -> CaughtUp` transition.
+
+# Anticipated Interface
+
+- Whenever necessary, the Diffusion Layer can ask the Consensus Layer for the ledger peer information.
+  EG `lpGetPeers :: STM m (SlotNo, [(PoolStake, NonEmpty RelayAccessPoint)])`.
```
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/docs/bootstrap-peers-IER.md]: I updated this section, please check

**Comment by amesgen** [comment]
> Could we put this in a ADR, [eg](https://github.com/input-output-hk/cardano-ledger/blob/master/docs/adr/2022-12-05_002-version-serialization-schemes.md). If so we might want to start off with [such a doc](https://github.com/input-output-hk/cardano-ledger/blob/master/docs/adr/2022-12-01_001-record-architectural-decisions.md).

We discussed this, and we will do this, but in a separate PR.

File: docs/website/docs/bootstrap-peers-IER.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/docs/bootstrap-peers-IER.md]: Rephrased this again after synchronous discussion with @nfrisby

File: docs/website/docs/bootstrap-peers-IER.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/docs/bootstrap-peers-IER.md]: We did chat about this synchronously, should now be resolved

File: docs/website/contents/for-developers/BootstrapPeersIER.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  This data will likely come from configuration files, and so doesn't require communication between the Diffusion Layer and the Consensus Layer.
+
+- When in `CaughtUp`, the Diffusion Layer needs to know the addresses and weights of the registered stake pool relays---aka the stake distribution among the ledger peers.
+  This information is in the ledger state, and so the Consensus Layer will need to provide to the Diffusion Layer continually.
+
+- For the sake of efficient reuse, the Diffusion Layer must manage the disconnections from all peers upon the `CaughtUp -> OnlyBootstrap` transition.
+  It would be wasteful for the Consensus Layer to achieve this less carefully, eg by simply killing those threads.
+
+  Therefore, the Consensus Layer needs to promptly notify the Diffusion Layer when the node should make the `CaughtUp -> OnlyBootstrap` transition.
+  Ideally, this happens immediately, so the implementation will likely involve an `STM` `TVar`.
```
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/contents/for-developers/BootstrapPeersIER.md:57]: This paragraph here is about `CaughtUp -> OnlyBootstrap` (quickly "getting rid" of non-bootstrap peers is important for security without an LoE-like mechanism), while the second bullet point in the second desideratum above is about `OnlyBootstrap -> CaughtUp` (no direct security concern here in promptly disconnecting from bootstrap peers).

[the remark this answers is not in the document]
Comment (nfrisby) [reply docs/website/contents/for-developers/BootstrapPeersIER.md:57]: Yep, I parsed it backwards.

File: docs/website/contents/for-developers/BootstrapPeersIER.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+- `CaughtUp` state - The peers are chosen according to the P2P design, including the _ledger peers_ etc.
+
+**Desideratum 2.**
+In particular, the transitions should happen promptly.
+
+- `CaughtUp -> OnlyBootstrap` should be prompt in order to minimize the duration that the node is exposed to untrusted peers (aka non-bootstrap peers) while its stale volatile tip is making it vulnerable.
+  Delays here would directly threaten the security of the node.
+
+- `OnlyBootstrap -> CaughtUp` should be prompt so that the centralized, relatively-few bootstrap peers are relieved of load as soon as possible.
```
[the remark this answers is not in the document]
Comment (amesgen) [reply docs/website/contents/for-developers/BootstrapPeersIER.md:36]: > Determining the exact load constraints for the bootstrap peers is not yet finalized.

**Review by amesgen — APPROVED** [review]
Approving as per https://github.com/input-output-hk/ouroboros-consensus/pull/200#pullrequestreview-1598209851

**Pull request merged**